### PR TITLE
LMS - vien

### PR DESCRIPTION
**Pull request by vyeung**

### CODE REVIEW

File: SmoothStackProject1/src/driver/Main.java
Comment (STOzaki) [inline SmoothStackProject1/src/driver/Main.java:19]: I like the way you divided up the menu. That is a good idea. So you do not just have a huge nested if else or switch statement!

File: SmoothStackProject1/src/daoImplementations/PublisherDaoImpl.java
Comment (STOzaki) [inline SmoothStackProject1/src/daoImplementations/PublisherDaoImpl.java:156]: Maybe check for the opposite and then do that
				if(Integer.parseInt(strArray[0]) != publisher.getPublisherId()) {

Comment (STOzaki) [reply SmoothStackProject1/src/daoImplementations/PublisherDaoImpl.java:156]: But do not forget to change the bottom part to convert this into just a if statement

File: SmoothStackProject1/src/daoImplementations/PublisherDaoImpl.java
Comment (STOzaki) [inline SmoothStackProject1/src/daoImplementations/PublisherDaoImpl.java:98]: It is probably too late for this, but in the future you may not want to have the scanner so low in your code. You want to have the menu be the only thing that communicates with the user

File: SmoothStackProject1/src/daoImplementations/AuthorDaoImpl.java
Comment (STOzaki) [inline SmoothStackProject1/src/daoImplementations/AuthorDaoImpl.java:91]: Having a temp file is a good idea.

**Review by STOzaki — APPROVED** [review]

File: SmoothStackProject1/src/daoImplementations/AuthorDaoImpl.java
Comment (kingjon3377) [inline SmoothStackProject1/src/daoImplementations/AuthorDaoImpl.java:15]: Do you really want anyone in this package to have access (including write access, since it isn't final) to this field? If I were you I'd make it `private final`.

File: SmoothStackProject1/src/daoImplementations/AuthorDaoImpl.java
Comment (kingjon3377) [inline SmoothStackProject1/src/daoImplementations/AuthorDaoImpl.java:21]: If you use try-with-resources, you don't have to explicitly close the writer, and it's guaranteed to be closed if an exception is thrown (your current code leaks the file handle in that case).

File: SmoothStackProject1/src/daoImplementations/AuthorDaoImpl.java
Comment (kingjon3377) [inline SmoothStackProject1/src/daoImplementations/AuthorDaoImpl.java:24]: You might want to wrap the FileWriter in a PrintWriter and use `printWriter.printf("%d,%s%n", author.getAuthorId(), author.getAuthorName());`

File: SmoothStackProject1/src/daoImplementations/AuthorDaoImpl.java
Comment (kingjon3377) [inline SmoothStackProject1/src/daoImplementations/AuthorDaoImpl.java:31]: Printing a stack trace gives a horrible UI experience; you should print a brief user-friendly message (to System.out or System.err if you don't take a stream for such UI messages as a constructor parameter) and log the exception using a logger, or simply declare this method as throwing `IOException` and handle it in a higher layer.

File: SmoothStackProject1/src/daoImplementations/AuthorDaoImpl.java
Comment (kingjon3377) [inline SmoothStackProject1/src/daoImplementations/AuthorDaoImpl.java:52]: If an exception occurs, a stack trace will appear on the console, but the caller will see the method succeed and receive a purely-default-state `Author`. If I were writing the calling code, I would want to handle the exception myself rather than getting an `Author` that didn't represent what I asked for.

File: SmoothStackProject1/src/daoImplementations/AuthorDaoImpl.java
Comment (kingjon3377) [inline SmoothStackProject1/src/daoImplementations/AuthorDaoImpl.java:65]: I like that you have column-title rows in your CSV files, though I wonder if you should perhaps check that this row you're skipping is in fact a title row and not real data (e.g. check if the field that will contain the ID says "ID" instead of a number).

File: SmoothStackProject1/src/daoImplementations/AuthorDaoImpl.java
Comment (kingjon3377) [inline SmoothStackProject1/src/daoImplementations/AuthorDaoImpl.java:73]: You should perhaps catch `NumberFormatException` somewhere, since that's what `Integer.parseInt` throws if the input is not numeric.

File: SmoothStackProject1/src/daoImplementations/AuthorDaoImpl.java
Comment (kingjon3377) [inline SmoothStackProject1/src/daoImplementations/AuthorDaoImpl.java:148]: You can avoid an empty 'if' block by inverting the condition.

File: SmoothStackProject1/src/daoInterface/AuthorDao.java
Comment (kingjon3377) [inline SmoothStackProject1/src/daoInterface/AuthorDao.java:8]: Because methods in an interface are public by default and cannot be anything else, you can omit the `public` modifier from them.

File: SmoothStackProject1/src/services/AuthorService.java
Comment (kingjon3377) [inline SmoothStackProject1/src/services/AuthorService.java:60]: The service classes are supposed to be a separate layer from the menu, right? So they shouldn't even know the menu exists, let alone call it. In this case, since this method will have (I presume) been called *from* the menu, just return.

**Review by kingjon3377 — CHANGES_REQUESTED** [review]
There are some design decisions you made that I really like (such as writing to a temporary file and having a header row), but there are also a couple of major architectural issues that I feel should block merging this code. The most major is the relationship (*recursion*) between the menu and the service code; the handling of exceptions in low-level methods by printing stack traces to the console and then returning as if successful is also problematic.

File: SmoothStackProject1/src/daoImplementations/AuthorDaoImpl.java
Comment (STOzaki) [reply SmoothStackProject1/src/daoImplementations/AuthorDaoImpl.java:148]: I agree

File: SmoothStackProject1/src/services/AuthorService.java
Comment (STOzaki) [reply SmoothStackProject1/src/services/AuthorService.java:60]: Yes, I agree

File: SmoothStackProject1/src/utils/Validate.java
Comment (STOzaki) [inline SmoothStackProject1/src/utils/Validate.java:18]: I do like the fact that you use regex to check if the phone number is in the right format

**Review by STOzaki — COMMENTED** [review]
You may want to add a few more tests for the service and/or the DAO

**Review by STOzaki — CHANGES_REQUESTED** [review]
Add a few more tests to test either service or DAO

File: SmoothStackProject1/src/driver/Main.java
Comment (STOzaki) [inline SmoothStackProject1/src/driver/Main.java:217]: Are you running a test here? If you so, then you may want to consider moving this to it's own section designated test/driver/MainTest.java

File: SmoothStackProject1/src/utils/Validate.java
Comment (kingjon3377) [reply SmoothStackProject1/src/utils/Validate.java:18]: Yes, if you're going to require phone numbers in a specific format, a regex is the right tool to use.

**Comment by al-amine** [comment]
I see that you haven't actually overridden the equals() and hashcode() in the models/pojos, which would be great if you do so.  because it's a good practice and it actually  helps a lot if you need to compare values...

**Pull request closed without merging**
